dstep/translator/Translator.d

```diff
@@ -69,48 +70,62 @@ class Translator
         write(outputFile, translateToString());
     }
 
+    static size_t countTrailingEndls(string a) 
+    {
+        size_t itr = a.length;
+
+        while (itr != 0 && a[itr - 1] == '\n')
+            --itr;
+                
+        return a.length - itr;
+    }
+
+    static size_t countLeadingEndls(string a)
+    {
+        size_t count = 0;
+
+        while (count < a.length && a[count] == '\n')
+            ++count;
+
+        return count;
+    }
+
+    static void endlAwareAppend(ref Appender!string a, string b)
+    {
+        import std.algorithm.comparison;
+
+        size_t aTrailing = countTrailingEndls(a.data);
+        size_t bLeading = countLeadingEndls(b);
+        size_t bTrailing = countTrailingEndls(b);
+
+        size_t num = min(bLeading + aTrailing, 2);
+        a.put("\n".replicate(num < aTrailing ? 0 : num - aTrailing));
+        a.put(b[bLeading..$-bTrailing]);
+        a.put("\n".replicate(min(bTrailing, 2)));
+    }
```

Comment:
They way I've implemented this in my [other project](https://github.com/d-widget-toolkit/jport/blob/dev/src/main/scala/dwt/jport/writers/NewlineWriter.scala) is that each node in the AST has a method that indicates if the node itself is multi-line or not. This is basically hard coded based on the type of the node. I also have information if there's a node coming after the current node at the same level (a next node). With this information it's trivial to handle the newlines: 

1. Add a newline after each node
2. If there's no next node we're done
3. If the current node or the next node is a multi-line node, add a newline

I also check the line numbers of the nodes to be able to group nodes in the output, but I don't think that's necessary for DStep, at least not for the first step.

I'll read through your comment tomorrow.